dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +241,260 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+                }
 
-                        throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
-                    }) };
+                // Register a sink for the application to receive foreground raw notifications
+                RegisterForegroundSinkHelper();
 
-                if (!isTaskRegistered)
+                // Register a PushNotificationBackgroundTask to handle background activation scenarios
+                if (!m_comActivatorRegistration)
                 {
-                    builder = BackgroundTaskBuilder();
-                    builder.Name(backgroundTaskFullName);
-
-                    PushNotificationTrigger trigger{};
-                    builder.SetTrigger(trigger);
+                    THROW_IF_FAILED(::CoRegisterClassObject(
+                        m_registeredClsid,
+                        winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                        CLSCTX_LOCAL_SERVER,
+                        REGCLS_MULTIPLEUSE,
+                        &m_comActivatorRegistration));
+                }
 
-                    THROW_HR_IF(E_NOTIMPL, !AppModel::Identity::IsPackagedProcess());
+                scopeExitToCleanRegistrations.release();
+            }
+            else
+            {
+                auto scopeExitToCleanRegistrations{ wil::scope_exit([&]()
+                {
+                    m_comActivatorRegistration.reset();
+                    m_protocolRegistration = false;
 
-                    // In case the interface is not supported, let it throw.
-                    auto builder5{ builder.as<winrt::IBackgroundTaskBuilder5>() };
-                    builder5.SetTaskEntryPointClsid(taskClsid);
-                    winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
-                    for (auto condition : conditions)
+                    auto notificationsLongRunningPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                    if (PushNotificationHelpers::IsPackagedAppScenario())
                     {
-                        builder.AddCondition(condition);
+                        auto appUserModelId{ PushNotificationHelpers::GetAppUserModelId() };
+                        PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get());
                     }
-                }
-            }
+                    else
+                    {
+                        notificationsLongRunningPlatform->UnregisterForegroundActivator(m_processName.get());
+                    }
+                    notificationsLongRunningPlatform->UnregisterFullTrustApplication(m_processName.get());
 
-            BackgroundTaskRegistration registeredTaskFromBuilder{ nullptr };
+                }
+                ) };
 
-            auto scopeExitToCleanRegistrations{ wil::scope_exit(
-                [&]()
+                if (!m_protocolRegistration)
```

Comment:
This is the wrong check. m_protocolRegistration is true for unpackaged apps only.